sdk/log: self observability: batch log processor metrics

mahendrabishnoi2 opened the pull request:
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/7017
Resolve #2547 

Adds support for below metrics for sdk/log#BatchProcessor
- `otel.sdk.processor.log.processed`
- `otel.sdk.processor.log.queue.size`
- `otel.sdk.processor.log.queue.capacity`

These metrics are defined at  https://github.com/open-telemetry/semantic-conventions/blob/v1.36.0/docs/otel/sdk-metrics.md and are experimental. Because of this, these use global MeterProvider and are behind a feature flag `OTEL_GO_X_OBSERVABILITY`.



<details>
<summary>Observability Implementation Checklist</summary>

## Observability Implementation Checklist
Based on the [project Observability guidelines](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/CONTRIBUTING.md#observability), ensure the following are completed:

### Environment Variable Activation
* [x] Observability features are disabled by default
* [x] Features are activated through the `OTEL_GO_X_OBSERVABILITY` environment variable
* [x] Use consistent pattern with `x.Observability.Enabled()` check [^1]
* [x] Follow established experimental feature pattern [^2][^3]

[^1]: https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L101-L103
[^2]: https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/x/x.go
[^3]: https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/sdk/internal/x/x.go

### Encapsulation
* [x] Instrumentation is encapsulated within a dedicated `struct` (e.g., [`Instrumentation`](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L86-L94))
* [x] Instrumentation is not mixed into the instrumented component
* [x] Instrumentation code is in its own file or package if complex/reused
* [x] Instrumentation setup doesn't bloat the main component code

### Initialization
* [x] Initialization is only done when observability is enabled
* [ ] Setup is explicit and side-effect free
* [ ] Return errors from initialization when appropriate
* [ ] Use the global Meter provider (e.g., `otel.GetMeterProvider()`)
* [ ] Include proper meter configuration with:
* [ ] Instrumentation package name is used for the Meter
* [ ] Instrumentation version (e.g. [`Version`](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L40-L43))
* [ ] Schema URL (e.g. [`SchemaURL`](https://github.com/open-telemetry/opentelemetry-go/blob/e4ab3141123d0811125a69823dbbe4d9ec5a9b8f/exporters/stdout/stdouttrace/internal/observ/instrumentation.go#L36-L38))

### Performance
* [ ] Little to no overhead when observability is disabled
* [ ] Expensive operations are only executed when observability is enabled
* [ ] When enabled, instrumentation code paths are optimized to reduce allocation/computation overhead

#### Attribute and Option Allocation Management
* [ ] Use `sync.Pool` for attribute slices and options with dynamic attributes
* [ ] Pool objects are properly reset before returning to pool
* [ ] Pools are scoped for maximum efficiency while ensuring correctness

#### Caching
* [ ] Static attribute sets known at compile time are pre-computed and cached
* [ ] Common attribute combinations use lookup tables/maps

#### Benchmarking
* [ ] Benchmarks provided for all instrumentation code
* [ ] Benchmark scenarios include both enabled and disabled observability
* [ ] Benchmark results show impact on allocs/op, B/op, and ns/op (use `b.ReportAllocs()` in benchmarks)

### Error Handling and Robustness
* [ ] Errors are reported back to caller when possible
* [ ] Partial failures are handled gracefully
* [ ] Use partially initialized components when available
* [ ] Return errors to caller instead of only using `otel.Handle()`
* [ ] Use `otel.Handle()` only when component cannot report error to user

### Context Propagation
* [ ] Observability measurements receive the context from the function being measured (don't break context propagation by using `context.Background()`)

### Semantic Conventions Compliance
* [ ] All metrics follow [OpenTelemetry Semantic Conventions](https://github.com/open-telemetry/semantic-conventions/blob/5ee549b1ce30fe11fcb9b7e3bd35ebfb363f467f/docs/otel/sdk-metrics.md)
* [ ] Use the [`otelconv`](https://pkg.go.dev/go.opentelemetry.io/otel@v1.38.0/semconv/v1.37.0/otelconv) convenience package for metric semantic conventions
* [ ] Component names follow semantic conventions
* [ ] Use package path scope type as stable identifier for non-standard components
* [ ] Component names are stable unique identifiers
* [ ] Use global counter for uniqueness if necessary
* [ ] Component ID counter is resettable for deterministic testing

### Testing
* [ ] Use deterministic testing with isolated state
* [ ] Restore previous state after tests (`t.Cleanup()`)
* [ ] Isolate meter provider for testing
* [ ] Use `t.Setenv()` for environment variable testing
* [ ] Reset component ID counter for deterministic component names
* [ ] Test order doesn't affect results
</details>

mahendrabishnoi2 committed:
- copied processorIDCounter
- setup and teardown for self observability. Includes setting up counters for queue capacity, size, log processed. Also includes callback registration for queue cap and size
- doc comments pending
- has some todo comments that I plan to address by discussion on PR review

mahendrabishnoi2 committed:
- added hook to call configureSelfObservability on BatchProcessor creation
- increment logProcessedCounter on push to exporter or drops from queue
- pending: flush methods where we are pushing to exporter

mahendrabishnoi2 committed:
- run `make precommit`

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
flatten setup for self-observability to make it transparent

mahendrabishnoi2 committed:
add metricsExporter, a wrapper to record successful log processed metric

mahendrabishnoi2 committed:
integrate with metricsExporter when self observability is enabled

mahendrabishnoi2 committed:
don't record metric when a log is added to queue, update comment to reflect what it means for a log to be processed

mahendrabishnoi2 committed:
update CHANGELOG.md and README.md (sdk/log/internal/x/README.md)

MrAlias commented:
@mahendrabishnoi2 I just wanted to check in with this PR. Are you still able to update this PR or open a new PR to address the instrumentation requirements?

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
self observability -> observability

mahendrabishnoi2 committed:
instrumentation implementation in a separate observ package as per new guidelines

mahendrabishnoi2 committed:
remove the wrapped exporter

mahendrabishnoi2 committed:
use newly created BLP abstraction for observability

mahendrabishnoi2 committed:
re-add metricsExporter with newly created struct (BLP) fo observability and integrate it with processor.

mahendrabishnoi2 committed:
use generated counter package for component names

mahendrabishnoi2 committed:
test cases for BLP

mahendrabishnoi2 committed:
make precommit

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
update CHANGELOG.md

mahendrabishnoi2 committed:
remove duplicate ScopeName

dashpole commented on sdk/log/internal/observ/batch_log_processor.go:
```diff
@@ -0,0 +1,121 @@
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/sdk"
+	"go.opentelemetry.io/otel/sdk/log/internal/x"
+	semconv "go.opentelemetry.io/otel/semconv/v1.37.0"
```
Needs to be updated to 1.40.0

dashpole commented on sdk/log/internal/observ/batch_log_processor.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (b *BLP) Processed(ctx context.Context, n int64) {
+	b.processed.Add(ctx, n, b.processedOpts...)
```
Check b.processed.Enabled(ctx) here to only call if the instrument is enabled.

mahendrabishnoi2 replied:
We have this block in `batch_log_processor.go`. This basically doesn't setup `metricsExporter` at all and these methods won't be called.
```
	if b.inst != nil {
		exporter = newMetricsExporter(exporter, b.inst)
	}
```

Let me know if I still need to do it?

mahendrabishnoi2 replied:
Added

dashpole commented on sdk/log/internal/observ/batch_log_processor.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (b *BLP) ProcessedQueueFull(ctx context.Context, n int64) {
+	b.processed.Add(ctx, n, b.processedQueueFullOpts...)
```
Check b.processed.Enabled(ctx) here to only call if the instrument is enabled.

dashpole reviewed (commented):
Not sure if you are still working on this, but this would be relevant if someone else picks this up.

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 commented:
> Not sure if you are still working on this, but this would be relevant if someone else picks this up.

I'll raise an updated PR today. Was just waiting for confirmation on my approach as I mentioned in issue as well.
Thanks @dashpole

mahendrabishnoi2 committed:
update semconv version

mahendrabishnoi2 committed:
check `b.processed.Enabled(ctx)` before updating metrics

mahendrabishnoi2 committed:
add tests

mahendrabishnoi2 committed:
fix CHANGELOG.md

mahendrabishnoi2 committed:
bypass lint for uint64 to int64 conversion as metric expects int64

mahendrabishnoi2 committed:
fix ci failures - add view to ignore batch log processor metrics

mahendrabishnoi2 committed:
make precommit

dashpole commented:
Please remove WIP from the title and mark it ready for review when it is ready for others to take a look

mahendrabishnoi2 committed:
minor refactoring in benchmark tests to reduce verbosity/duplicate code

mahendrabishnoi2 committed:
minor refactoring in benchmark tests to reduce verbosity/duplicate code

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 commented:
```
goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/log/internal/observ
cpu: Apple M1 Pro
                         │ blp_bench.txt │
                         │    sec/op     │
BLP/Processed-8             0.4520n ± 3%
BLP/ProcessedQueueFull-8    0.4463n ± 1%
BLP/Callback-8               2.692µ ± 8%
geomean                      8.159n

                         │ blp_bench.txt  │
                         │      B/op      │
BLP/Processed-8              0.000 ± 0%
BLP/ProcessedQueueFull-8     0.000 ± 0%
BLP/Callback-8             2.814Ki ± 0%
geomean                                 ¹
¹ summaries must be >0 to compute geomean

                         │ blp_bench.txt │
                         │   allocs/op   │
BLP/Processed-8             0.000 ± 0%
BLP/ProcessedQueueFull-8    0.000 ± 0%
BLP/Callback-8              26.00 ± 0%
geomean                                ¹
¹ summaries must be >0 to compute geomean

```

dashpole commented on CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add experimental support for splitting metric data across multiple batches in `go.opentelemetry.io/otel/sdk/metric`.
   Set `OTEL_GO_X_METRIC_EXPORT_BATCH_SIZE=<max_size>` to enable for all periodic readers.
   See `go.opentelemetry.io/otel/sdk/metric/internal/x` for feature documentation. (#8071)
+- Add experimental observability metrics in `go.opentelemetry.io/otel/sdk/log`. (#7124)
```
```suggestion
- Add experimental observability metrics to the BatchProcessor in `go.opentelemetry.io/otel/sdk/log`. (#7124)
```

pellared commented on sdk/log/internal/observ/batch_log_processor.go:
```diff
@@ -0,0 +1,125 @@
+	}, err
+}
+
+func (b *BLP) Shutdown() error {
```
Here is how Codex found the same issue:

If `meter.RegisterCallback` returns an error, reg is left nil here but `NewBLP` still returns a non-nil `*BLP`, so `NewBatchProcessor` installs it and later `(*BLP).Shutdown` calls `b.reg.Unregister()` on a nil registration. In any process that enables `OTEL_GO_X_OBSERVABILITY` with a custom or partially configured `MeterProvider`, shutdown turns a handled instrumentation error into a panic.

mahendrabishnoi2 replied:
I have not added `b == nil` check as this is internal api and on all callsites, we already check for `b.inst == nil`.
Do let me know if this check also needs to be added.

pellared replied:
I think it is safer to add this check (defensive programming)

mahendrabishnoi2 replied:
done. Added `b == nil || b.reg == nil` in if condition
https://github.com/open-telemetry/opentelemetry-go/pull/7124/changes/b3019d523ca6025b9dac31195b3803a625fa2edf

pellared commented on sdk/log/batch.go:
```diff
@@ -225,6 +255,9 @@ func (b *BatchProcessor) Shutdown(ctx context.Context) error {
 	// Flush remaining queued before exporter shutdown.
 	err := b.exporter.Export(ctx, b.q.Flush())
+	if b.inst != nil {
+		err = errors.Join(err, b.inst.Shutdown())
```
When records have overflowed the ring buffer since the last poll iteration, `Shutdown` closes `pollKill` and can exit the poll goroutine without ever consuming `q.Dropped()`. Because `ProcessedQueueFull` is only driven from that goroutine, an application that shuts down under backpressure can silently miss `otel.sdk.processor.log.processed{error.type="queue_full"}` increments for the last dropped log records, so the new observability metric under-reports exactly the shutdown path users care about most.

I think that addressing this would be very hard and I would rather create an issue for tracking this.

mahendrabishnoi2 replied:
https://github.com/open-telemetry/opentelemetry-go/issues/8213

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
address review comments: fix nil reg panic in Shutdown, changelog wording

mahendrabishnoi2 commented on sdk/log/batch_test.go:
```diff
@@ -673,3 +683,199 @@ func BenchmarkBatchProcessorOnEmit(b *testing.B) {
+
+func TestBatchProcessorMetricsDisabled(t *testing.T) {
+	t.Setenv("OTEL_GO_X_OBSERVABILITY", "false")
+
```
similar pattern as other package tests, setting to 0

pellared replied:
Can we used this pattern https://github.com/open-telemetry/opentelemetry-go/blob/9a68034bd45c6f24c481d9f9c87ebbee0a61482f/exporters/otlp/otlpmetric/otlpmetrichttp/client_test.go#L547-L550 ?

mahendrabishnoi2 replied:
done in https://github.com/open-telemetry/opentelemetry-go/pull/7124/changes/c7e2ffb268104c4cbe264cdbdfd9a0815a896981

mahendrabishnoi2 commented on sdk/log/batch_test.go:
```diff
@@ -673,3 +683,199 @@ func BenchmarkBatchProcessorOnEmit(b *testing.B) {
+	e.Stop()
+	require.NoError(t, bp.Shutdown(ctx))
+}
+
+func TestBatchProcessorMetrics(t *testing.T) {
+	counter.SetExporterID(blpComponentID)
+
+	t.Setenv("OTEL_GO_X_OBSERVABILITY", "true")
```
same as above

pellared replied:
https://github.com/open-telemetry/opentelemetry-go/pull/7124#discussion_r3304691912

mahendrabishnoi2 replied:
done in https://github.com/open-telemetry/opentelemetry-go/pull/7124/changes/c7e2ffb268104c4cbe264cdbdfd9a0815a896981

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
fix github checks

mahendrabishnoi2 commented:
One lint failure. Unrelated to code changes of this PR.
```
### Errors in CHANGELOG.md

* [404] <https://github.com/open-telemetry/opentelemetry-go/security/advisories/GHSA-995v-fvrw-c78m> | Rejected status code: 404 Not Found (configurable with "accept" option)


Notice: Summary report available at: https://github.com/open-telemetry/opentelemetry-go/actions/runs/26398659775#summary-77705321176
```

mahendrabishnoi2 commented:
Hi @pellared, Please take a look if your comments have been addressed. Thanks

pellared committed:
Merge branch 'main' into logs-batch-processor-metrics

pellared commented on sdk/log/batch.go:
```diff
@@ -111,6 +117,31 @@ func NewBatchProcessor(exporter Exporter, opts ...BatchProcessorOption) *BatchPr
+		pollKill:    make(chan struct{}),
+	}
+
+	var err error
+	b.inst, err = observ.NewBLP(
+		counter.NextExporterID(),
+		func() int64 { return int64(b.q.Len()) },
+		int64(cfg.maxQSize.Value),
+	)
```
Has this comment been addressed?

mahendrabishnoi2 replied:
My bad. I missed replying here. 
I decided not to follow this suggestion as we use this exact pattern in multiple places within this codebase. Examples: [search: counter.NextExporterID()](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-go%20counter.NextExporterID()&type=code), [code-ref](https://github.com/open-telemetry/opentelemetry-go/blob/6e2b9214f05e4d42a5215acafc45a0055c0e4513/exporters/stdout/stdouttrace/trace.go#L39)

pellared commented on sdk/log/batch.go:
```diff
@@ -111,6 +117,31 @@ func NewBatchProcessor(exporter Exporter, opts ...BatchProcessorOption) *BatchPr
+		pollKill:    make(chan struct{}),
+	}
+
+	var err error
+	b.inst, err = observ.NewBLP(
+		counter.NextExporterID(),
+		func() int64 { return int64(b.q.Len()) },
+		int64(cfg.maxQSize.Value),
+	)
+	if err != nil {
+		otel.Handle(err)
+	}
+
+	// Wrap exporter with metrics recording if observability is enabled.
+	// This must be the innermost wrapper (closest to user exporter) to record
+	// metrics just before calling the actual exporter.
+	if b.inst != nil {
+		exporter = newMetricsExporter(exporter, b.inst)
+	}
```
Has this comment been addressed?

mahendrabishnoi2 replied:
Addressed in https://github.com/open-telemetry/opentelemetry-go/pull/7124/changes/a917056807a31ae2092f5815953f023ce29191c4

Previously I didn't reply as this was comment from copilot and my pattern was similar to this from batch_span_processor https://github.com/open-telemetry/opentelemetry-go/blob/6e2b9214f05e4d42a5215acafc45a0055c0e4513/sdk/trace/internal/observ/batch_span_processor.go#L50-L109

pellared commented:
> Hi @pellared, Please take a look if your comments have been addressed. Thanks

@mahendrabishnoi2, it looks that there are still comments that need to be addressed. Please double-check.

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
readd in CHANGELOG.md after main conflict resolve due to new release

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
address review comment: nil check before blp shutdown

mahendrabishnoi2 committed:
address review comment: reset counter with original value in test

mahendrabishnoi2 committed:
address review comment: NewBLP can return non nil for both BLP and error

mahendrabishnoi2 committed:
fix CHANGELOG.md

mahendrabishnoi2 committed:
make precommit

mahendrabishnoi2 committed:
address copilot comments

mahendrabishnoi2 commented on sdk/log/batch_test.go:
```diff
@@ -673,3 +683,203 @@ func BenchmarkBatchProcessorOnEmit(b *testing.B) {
+	r := new(Record)
+	r.SetBody(log.BoolValue(true))
+	require.NoError(t, bp.OnEmit(ctx, r))
+	require.Eventually(t, func() bool {
+		return e.ExportN() > 0
+	}, 2*time.Second, time.Microsecond, "export not started")
```
followed existing pattern 
https://github.com/open-telemetry/opentelemetry-go/blob/dcfc16cebd0674eaaa7fe5316be8abe8a0f6d35e/sdk/log/exporter_test.go#L442-L444

happy to change to higher value for ticker (time.Microsecond) if needed. Haven't noticed flakiness because of this. 

Same for next 2 comments from copilot.

mahendrabishnoi2 commented on sdk/log/exporter.go:
```diff
@@ -324,3 +325,26 @@ func (e *bufferExporter) Shutdown(ctx context.Context) error {
+}
+
+// Export records the number of log records as a metric then forwards
+// them to the wrapped Exporter. Error returned from wrapped exporter
+// is not considered as per specification (to be measured by exporter).
+func (e *metricsExporter) Export(ctx context.Context, records []Record) error {
+	e.inst.Processed(ctx, int64(len(records)))
+	return e.Exporter.Export(ctx, records)
+}
```
addressed in https://github.com/open-telemetry/opentelemetry-go/pull/7124/commits/96bfdcb5e450529e901033c0363651411d850cb6

MrAlias replied:
This still needs one more tweak before resolving. The current guard avoids the nil dereference, but it returns before calling the wrapped exporter when `e.inst == nil`, so a misconstructed wrapper would silently drop records. It should skip only the `Processed` call and still return `e.Exporter.Export(ctx, records)`.

mahendrabishnoi2 replied:
addressed in https://github.com/open-telemetry/opentelemetry-go/pull/7124/changes/27c8cd5a23484a52b53d00bdab438068ade5a6ee

mahendrabishnoi2 commented on sdk/log/internal/observ/batch_log_processor.go:
```diff
@@ -0,0 +1,122 @@
+
+	// Register callback for async metrics
+	obsOpts := []metric.ObserveOption{metric.WithAttributeSet(set)}
+	reg, err := meter.RegisterCallback(
+		func(_ context.Context, o metric.Observer) error {
+			o.ObserveInt64(qSizeInst, qLen(), obsOpts...)
+			o.ObserveInt64(qCapInst, qMax, obsOpts...)
+			return nil
+		},
+		qSizeInst,
+		qCapInst,
+	)
```
addressed in https://github.com/open-telemetry/opentelemetry-go/pull/7124/commits/96bfdcb5e450529e901033c0363651411d850cb6

mahendrabishnoi2 commented on sdk/log/batch_test.go:
```diff
@@ -673,3 +683,203 @@ func BenchmarkBatchProcessorOnEmit(b *testing.B) {
+	t.Cleanup(func() { global.SetLogger(origLogger) })
+	buf := new(concurrentBuffer)
+	stdr.SetVerbosity(1)
+	global.SetLogger(stdr.New(stdlog.New(buf, "", 0)))
```
addressed in https://github.com/open-telemetry/opentelemetry-go/pull/7124/commits/96bfdcb5e450529e901033c0363651411d850cb6

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
fix lint failure

MrAlias commented on sdk/log/batch_test.go:
```diff
@@ -673,3 +683,203 @@ func BenchmarkBatchProcessorOnEmit(b *testing.B) {
+	)
+
+	require.NoError(t, bp.OnEmit(ctx, r))
+	require.Eventually(t, func() bool {
+		return len(bp.exporter.input) == cap(bp.exporter.input)
+	}, 2*time.Second, time.Microsecond, "buffer channel not filled")
```
Covered by the author's reply on the first matching polling-interval thread: this follows existing `sdk/log` test usage of `time.Microsecond`. Treating this as non-blocking automation feedback.

MrAlias commented on sdk/log/batch_test.go:
```diff
@@ -673,3 +683,203 @@ func BenchmarkBatchProcessorOnEmit(b *testing.B) {
+	require.NoError(t, bp.OnEmit(ctx, r))
+
+	wantMsg := `"level"=1 "msg"="dropped log records" "dropped"=1`
+	require.Eventually(t, func() bool {
+		return strings.Contains(buf.String(), wantMsg)
+	}, 2*time.Second, time.Microsecond, "drop not detected")
```
Covered by the author's reply on the first matching polling-interval thread: this follows existing `sdk/log` test usage of `time.Microsecond`. Treating this as non-blocking automation feedback.

mahendrabishnoi2 committed:
review pr comment: always call wrapper exporter, only skip Processed
call on inst when its nil

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

pellared commented:
Can you please update the PR. Sorry for inconvenience.

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
fix issue from emrge conflict

mahendrabishnoi2 committed:
fix ci issues

mahendrabishnoi2 committed:
fix CHANGELOG.md: revert unintended formatting changes from merge conflict resolution

mahendrabishnoi2 committed:
semconv 1.42

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

pellared commented:
@mahendrabishnoi2, can you address Copilot comments?

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
copilt review comments:
1. clamp dropped count to math.MaxInt64 to prevent overflow - defensive
programming
2. call Shutdown() on b.inst for ctx.Done() path

mahendrabishnoi2 commented on sdk/log/batch.go:
```diff
@@ -153,6 +180,9 @@ func (b *BatchProcessor) poll(interval time.Duration) (done chan struct{}) {
 			}
 
 			if d := b.q.Dropped(); d > 0 {
+				if b.inst != nil {
+					b.inst.ProcessedQueueFull(ctx, int64(d)) //nolint: gosec
+				}
 				global.Warn("dropped log records", "dropped", d)
 			}
```
clamped to math.MaxInt64

I still had to keep `//nolint: gosec` as linter was still complaining

mahendrabishnoi2 commented on sdk/log/batch.go:
```diff
@@ -225,6 +255,9 @@ func (b *BatchProcessor) Shutdown(ctx context.Context) error {
+	if b.inst != nil {
+		err = errors.Join(err, b.inst.Shutdown())
+	}
 	return errors.Join(err, b.exporter.Shutdown(ctx))
```
b.inst.Shutdown() added for ctx.Done() early return path

mahendrabishnoi2 committed:
copilot review: unregister callback

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 committed:
Merge branch 'main' into logs-batch-processor-metrics

mahendrabishnoi2 commented:
> @mahendrabishnoi2, can you address Copilot comments?

All issues fixed, no comments on re-review from CoPilot. 
Please let me know if anything else needs to be done here. Thanks.

mahendrabishnoi2 commented:
In previous CI check, govuln check failed due to transient network error (`govulncheck: fetching vulnerabilities: Get "https://vuln.go.dev/index/modules.json.gz": read tcp 10.1.1.172:35400->34.117.213.18:443: read: connection reset by peer
`). This should be resolved on this rerun.

MrAlias reviewed (approved):
The batch processor metrics implementation looks good to me. I plan to follow up separately to add `error.type="already_shutdown"` measurement for records dropped after processor shutdown.

MrAlias committed:
Merge branch 'main' into logs-batch-processor-metrics

MrAlias committed:
Fix lint

MrAlias committed:
Merge branch 'main' into logs-batch-processor-metrics

MrAlias committed:
Merge main into logs batch processor metrics
